Sources/Workspace/DefaultPluginScriptRunner.swift

```diff
@@ -125,204 +127,257 @@ public struct DefaultPluginScriptRunner: PluginScriptRunner, Cancellable {
+        let compilerInputHash: String?
+        do {
+            // Include the full compiler arguments and environment, and the contents of the source files.
+            let stream = BufferedOutputByteStream()
+            stream <<< commandLine
+            for (key, value) in toolchain.swiftCompilerEnvironment.sorted(by: { $0.key < $1.key }) {
+                stream <<< "\(key)=\(value)\n"
+            }
+            for sourceFile in sourceFiles {
+                try stream <<< fileSystem.readFileContents(sourceFile).contents
+            }
+            compilerInputHash = stream.bytes.sha256Checksum
+            observabilityScope.emit(debug: "Computed hash of plugin compilation inputs: \(compilerInputHash!)")
+        }
+        catch {
+            // We couldn't compute the hash. We warn about it but proceed with the compilation (a cache miss).
+            observabilityScope.emit(debug: "... couldn't compute hash of plugin compilation inputs (\(error))")
+            compilerInputHash = .none
+        }
         
-            // Hash the command line and the contents of the source files to decide whether we need to recompile the plugin executable.
-            let compilerInputsHash: String?
-            do {
-                // We include the full command line, the environment, and the contents of the source files.
-                let stream = BufferedOutputByteStream()
-                stream <<< command
-                for (key, value) in toolchain.swiftCompilerEnvironment.sorted(by: { $0.key < $1.key }) {
-                    stream <<< "\(key)=\(value)\n"
+        /// Persisted information about the last time the compiler was invoked.
+        struct PersistedCompilationState: Codable {
+            var commandLine: [String]
+            var environment: [String:String]
+            var inputHash: String?
+            var duration: Double
+            var output: String
+            var result: Result
+            enum Result: Equatable, Codable {
+                case exit(code: Int32)
+                case signal(number: Int32)
+                
+                init(_ processExitStatus: ProcessResult.ExitStatus) {
+                    switch processExitStatus {
+                    case .terminated(let code):
+                        self = .exit(code: code)
+                    case .signalled(let signal):
+                        self = .signal(number: signal)
+                    }
                 }
-                for sourceFile in sources.paths {
-                    try stream <<< fileSystem.readFileContents(sourceFile).contents
+            }
+            
+            var succeeded: Bool {
+                return result == .exit(code: 0)
+            }
+        }
+        
+        // Check if we already have a compiled executable and a persisted state (we only recompile if things have changed).
+        let stateFilePath = self.cacheDir.appending(component: execName + ".state")
```

Comment:
nit: use "json" as extension, and names the file "state"

Reply:
Just the literal name "state" and not based on the name of the executable?  In that case we should maybe call the executable just "plugin".  But that makes it harder to see which plugin it is when looking for example using `top`.

I was trying to keep all three files (executable, state, and diagnostics) with the same base name derived from the plugin name.  So maybe just changing it to "json" here makes sense.

Reply:
sry I meant `execName + "-state.json"`

Reply:
That makes sense.